vllm/multimodal/utils.py

```diff
@@ -0,0 +1,71 @@
+import base64
+from io import BytesIO
+from typing import Union
+
+import requests
+from PIL import Image
+
+from vllm.config import ModelConfig
+from vllm.envs import VLLM_IMAGE_FETCH_TIMEOUT
+from vllm.multimodal.image import ImagePixelData
+from vllm.utils import make_async
+
+
+def load_image_from_base64(image: Union[bytes, str]) -> Image.Image:
+    """Load image from base64 format"""
+    return Image.open(BytesIO(base64.b64decode(image)))
+
+
+def encode_image_base64(image: Image.Image, format: str = 'JPEG') -> str:
+    """encode image to base64 format."""
+
+    buffered = BytesIO()
+    if format == 'JPEG':
+        image = image.convert('RGB')
+    image.save(buffered, format)
+    return base64.b64encode(buffered.getvalue()).decode('utf-8')
+
+
+def fetch_image(image_url: str) -> Image.Image:
+    """Load image from a url or base64 encoded openai GPT4V format"""
+
+    # Avoid circular import
+    from vllm import __version__ as VLLM_VERSION
+
+    headers = {"User-Agent": f"vLLM/{VLLM_VERSION}"}
+    if image_url.startswith('http'):
+        response = requests.get(image_url,
+                                headers=headers,
+                                timeout=VLLM_IMAGE_FETCH_TIMEOUT)
+        response.raise_for_status()
+
+        # Open the image using PIL
+        img = Image.open(BytesIO(response.content))
+    elif image_url.startswith('data:image'):
+        img = load_image_from_base64(image_url.split(',')[1])
+    else:
+        raise ValueError("Invalid image url: A valid image url must start "
+                         "with either 'data:image' or 'http'.")
+
+    return img
+
+
+async_fetch_image = make_async(fetch_image)
```

Comment:
FYI the perf will be really bad here because it will be spawning a thread per image download (won't block the event loop but will consume Python GIL). It might be easier to introduce dependency for aiohttp or httpx and make it natively async so it's faster to benchmark.

Reply:
Sounds good, I will take a look and modify this.